### PR TITLE
Add unit tests for Session

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Added disambiguating text to more debug statements.
Fixed a race condition when concurrently creating sessions.
Added support for sender attach performative and encoding SASL frames.
Fixed a hang in session mux for malformed attach frame.
Added some clarifying comments to a few chunks of code.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -476,6 +476,16 @@ func (c *conn) mux() {
 				session.remoteChannel = fr.Channel
 				sessionsByRemoteChannel[fr.Channel] = session
 
+			case *frames.PerformEnd:
```
Comment (jhendrixMSFT) [inline conn.go:479]: Fix for concurrent session bug.

File: conn.go [hunk 2]
```diff
@@ -476,6 +476,16 @@ func (c *conn) mux() {
...
+				session, ok = sessionsByRemoteChannel[fr.Channel]
+				if !ok {
+					c.err = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel", fr.Body, fr.Channel)
+				}
+				// we MUST remove the remote channel from our map as soon as we receive
+				// the ack (i.e. before passing it on to the session mux) on the session
+				// ending since the numbers are recycled.
+				delete(sessionsByRemoteChannel, fr.Channel)
```
Comment (richardpark-msft) [inline conn.go:487]: Can all the session deletion code just go here now?

```go
// session deletion
case s := <-c.DelSession:
 delete(sessionsByChannel, s.channel)
 channels.Remove(uint32(s.channel))
```

File: link_test.go [hunk 3]
```diff
@@ -316,3 +317,50 @@ func TestSourceName(t *testing.T) {
 		t.Errorf("Link Source Name does not match expected: %v got: %v", expectedSourceName, got.Key.name)
 	}
 }
+
+func TestSessionFlowDisablesTransfer(t *testing.T) {
+	t.Skip("TODO: finish for link testing")
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply link_test.go:322]: The test itself is incomplete, will wrap it up as part of link tests (I got halfway writing it when I realized it wasn't specific to sessions).

File: conn.go [hunk 4]
```diff
@@ -476,6 +476,16 @@ func (c *conn) mux() {
...
+
 			default:
 				session, ok = sessionsByRemoteChannel[fr.Channel]
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply conn.go:490]: Yes, any performative not handled here is passed on to the session.

File: conn.go [hunk 5]
```diff
@@ -476,6 +476,16 @@ func (c *conn) mux() {
 				session.remoteChannel = fr.Channel
 				sessionsByRemoteChannel[fr.Channel] = session
 
+			case *frames.PerformEnd:
+				session, ok = sessionsByRemoteChannel[fr.Channel]
+				if !ok {
+					c.err = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel", fr.Body, fr.Channel)
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply conn.go:482]: The end performative only has an error, so dumping it out I don't think is all that useful.  This was one of those "impossible-to-happen" negative tests I added, where if it _were_ to happen the mux would hang.  I did realize this if block is missing a break statement though (not that its absence is bad as the end-result is the same, but it does make the intention clear).

File: integration_test.go [hunk 6]
```diff
@@ -441,7 +441,7 @@ func TestIntegrationSessionHandleMax(t *testing.T) {
 		{
 			maxLinks: 4,
 			links:    5,
-			error:    regexp.MustCompile(`handle max \(3\)`),
+			error:    regexp.MustCompile(`handle max \(4\)`),
```
[the remark this answers is not in the document]
Comment (richardpark-msft) [reply integration_test.go:444]: Makes sense.

File: conn.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (richardpark-msft) [reply conn.go:482]: Makes sense.

File: conn.go (hunk 2 above)
Comment (richardpark-msft) [reply conn.go:488]: Ah, okay. Perhaps in the future, not sure if it's worth doing at all.

**Pull request merged**